Api client list accessors

ptzianos opened the pull request:
Fixes: #10463

Changes:
- Add a function to the API client so that a user can list the token accessors of the system
- Split and re-use function for parsing token accessors
- Add a small helper to decode temp root tokens

ptzianos commented:
note for reviewer: If the inclusion of the root token decoding function is a problem I have no problem splitting it away in a separate PR.

ncabatoff commented on api/secret.go:
```diff
@@ -113,69 +146,19 @@ func (s *Secret) TokenPolicies() ([]string, error) {
 	}
 
-	if s.Data == nil || s.Data["policies"] == nil {
+	if s.Data == nil {
```
Looks like you could get rid of the `if` altogether, since the test is already present in extractListOfStrings.

ptzianos replied:
Done

ncabatoff commented on api/secret.go:
```diff
@@ -113,69 +146,19 @@ func (s *Secret) TokenPolicies() ([]string, error) {
-		}
+	tokenPolicies, err := s.extractListOfStrings("policies")
+	if tokenPolicies == nil || err != nil {
+		return tokenPolicies, err
```
There are cases in the current code where we goto TOKEN_DONE despite having an empty tokenPolicies.  I think this `if`'s condition should be just `err != nil`.

ptzianos replied:
Yeah, I didn't notice it. Now the latest version of the code should be equivalent in functionality

ncabatoff commented on api/secret.go:
```diff
@@ -189,6 +172,23 @@ DONE:
+	}
+
+	if s.Auth != nil && len(s.Auth.Policies) > 0 {
+		return s.Auth.Policies, nil
```
This doesn't look correct.  Policies aren't accessors, right?

ptzianos replied:
Good catch. fixed now

ncabatoff commented on api/secret.go:
```diff
@@ -189,6 +172,23 @@ DONE:
 }
 
+// TokenAccessors returns the standardized list of accessors for the given secret
+func (s *Secret) TokenAccessors() ([]string, error) {
```
You name this `TokenAccessors`, but it could just as well be called `Keys` or something, right?  It's the same code that would be used to extract the result of any `list` operation from a `Secret`.

ptzianos replied:
Yes, you are right on that. I removed the if statement that would return the secret's accessor value because that was wrong. Instead the function call basically just invokes the `extractListOfStrings` with the correct parameters.

ptzianos commented:
@ncabatoff  it looks like I need some sort of authorization from someone in order to get my PR deployed on Vercel? Other than that I fixed all your suggestions and added tests which are passing
